Formula/grafana-cloud-agent.rb

```diff
@@ -0,0 +1,103 @@
+  sha256 "3b8cb72793888f676fb205ced065da173429289dd29c77b79b095350ec28bf35"
+  license "Apache-2.0"
+
+  livecheck do
+    url "https://github.com/grafana/agent/releases/latest"
+    regex(%r{href=.*?/tag/v?(\d+(?:\.\d+)+)["' >]}i)
+  end
+
+  depends_on "go" => :build
+
+  def install
+    mkdir_p buildpath/"src/github.com/grafana"
+    ln_sf buildpath, buildpath/"src/github.com/grafana/agent"
+
+    system "make", "agent"
+
+    bin.install "./cmd/agent/agent"
+    mv "#{bin}/agent", "#{bin}/grafana-cloud-agent"
+
+    (bin/"grafana-cloud-agent_brew_services").write <<~EOS
+      #!/bin/bash
+      exec #{bin}/grafana-cloud-agent $(<#{etc}/grafana-cloud-agent.args)
+    EOS
+
+    (buildpath/"grafana-cloud-agent.args").write <<~EOS
+      --config.file #{etc}/grafana-cloud-agent.yml
+    EOS
+
+    (buildpath/"grafana-cloud-agent.yml").write <<~EOS
+      server:
+        log_level: info
+        http_listen_port: 12345
+
+      prometheus:
+        wal_directory: /tmp/agent
+        global:
+          scrape_interval: 15s
+
+      integrations:
+        node_exporter:
+          enabled: true
+          rootfs_path: /host/root
+          sysfs_path: /host/sys
+          procfs_path: /host/proc
+        prometheus_remote_write:
+          - url: https://prometheus-us-central1.grafana.net/api/prom/push
+            basic_auth:
+              username: USERNAME
+              password: PASSWORD
+    EOS
+
+    etc.install "grafana-cloud-agent.args", "grafana-cloud-agent.yml"
+  end
+
+  def caveats
+    <<~EOS
+      The agent uses a configuration file that you must customize before running:
+        #{etc}/grafana-cloud-agent.yml
+
+      You can find the Configuration file reference at:
+        https://github.com/grafana/agent/blob/master/docs/configuration-reference.md
+
+      When run from `brew services`, `grafana-cloud-agent` is run from
+      `grafana-cloud-agent_brew_services` and uses the flags in:
+         #{etc}/grafana-cloud-agent.args
+    EOS
+  end
+
+  plist_options manual: "grafana-cloud-agent -config.file=#{HOMEBREW_PREFIX}/etc/grafana-cloud-agent.yml"
+
+  def plist
+    <<~EOS
+      <?xml version="1.0" encoding="UTF-8"?>
+        <!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+        <plist version="1.0">
+        <dict>
+          <key>Label</key>
+          <string>#{plist_name}</string>
+          <key>ProgramArguments</key>
+          <array>
+            <string>#{opt_bin}/grafana-cloud-agent_brew_services</string>
+          </array>
+          <key>RunAtLoad</key>
+          <true/>
+          <key>KeepAlive</key>
+          <false/>
+          <key>StandardErrorPath</key>
+          <string>#{var}/log/grafana-cloud-agent.err.log</string>
+          <key>StandardOutPath</key>
+          <string>#{var}/log/grafana-cloud-agent.log</string>
+        </dict>
+      </plist>
+    EOS
+  end
+
+  test do
+    system "#{bin}/grafana-cloud-agent", "-version"
```

Comment:
We need a test that exercises the some of the functionality of the app. Version checks or usage checks (foo --version or foo --help) are not sufficient, as explained in the formula cookbook.

In most cases, a good test would involve running a simple test case: run #{bin}/foo input.txt.

- Then you can check that the output is as expected (with assert_equal or assert_match on the output of shell_output)
- You can also check that an output file was created, if that is expected: assert_predicate testpath/"output.txt", :exist?

Some advice for specific cases:

- If the formula is a library, compile and run some simple code that links against it. It could be taken from upstream's documentation / source examples.
- If the formula is for a GUI program, try to find some function that runs as command-line only, like a format conversion, reading or displaying a config file, etc.
- If the software cannot function without credentials, a test could be to try to connect with invalid credentials (or without credentials) and confirm that it fails as expected.
- Same if the software requires a virtual machine, docker instance, etc. to be running.

Reply:
This formula is for an agent that is meant to be running as a long-lived process. There is no functionality to print back config either; the only thing I found that could work is running it with a non-existant config file, which will crash with an error message. Would that suffice?

Reply:
Things like `httpd` are also long-lived processes. You can check the test there.

Reply:
Thanks, that's a good pointer, I'll rewrite the test to exercise the built-in http server that serves metrics.